rpc/namenode.go

```diff
@@ -55,45 +67,96 @@ func (err *NamenodeError) Error() string {
 	return s
 }
 
-// NewNamenodeConnection creates a new connection to a Namenode, and preforms an
+type host struct {
+	address     string
+	lastFailure time.Time
+}
+
+// NewNamenodeConnection creates a new connection to a namenode and performs an
 // initial handshake.
 //
 // You probably want to use hdfs.New instead, which provides a higher-level
 // interface.
-func NewNamenodeConnection(address, user string) (*NamenodeConnection, error) {
-	conn, err := net.DialTimeout("tcp", address, connectTimeout)
-	if err != nil {
-		return nil, err
-	}
-
-	return WrapNamenodeConnection(conn, user)
+func NewNamenodeConnection(address string, user string) (*NamenodeConnection, error) {
+	return NewNamenodeConnectionWithOptions(NamenodeConnectionOptions{
+		Addresses: []string{address},
+		User:      user,
+	})
 }
 
-// WrapNamenodeConnection wraps an existing net.Conn to a Namenode, and preforms
-// an initial handshake.
-//
-// You probably want to use hdfs.New instead, which provides a higher-level
-// interface.
-func WrapNamenodeConnection(conn net.Conn, user string) (*NamenodeConnection, error) {
+// NewNamenodeConnectionWithOptions creates a new connection to a namenode with
+// the given options and performs an initial handshake.
+func NewNamenodeConnectionWithOptions(options NamenodeConnectionOptions) (*NamenodeConnection, error) {
+	// Build the list of hosts to be used for failover.
+	hostList := make([]*host, len(options.Addresses))
+	for i, addr := range options.Addresses {
+		hostList[i] = &host{address: addr}
+	}
+
 	// The ClientID is reused here both in the RPC headers (which requires a
 	// "globally unique" ID) and as the "client name" in various requests.
 	clientId := newClientID()
 	c := &NamenodeConnection{
 		clientId:   clientId,
 		clientName: "go-hdfs-" + string(clientId),
-		user:       user,
-		conn:       conn,
+		user:       options.User,
+		hostList:   hostList,
 	}
 
-	err := c.writeNamenodeHandshake()
+	err := c.resolveConnection()
 	if err != nil {
-		conn.Close()
-		return nil, fmt.Errorf("Error performing handshake: %s", err)
+		return nil, err
 	}
 
 	return c, nil
 }
 
+func (c *NamenodeConnection) resolveConnection() error {
+	if c.conn != nil {
+		return nil
+	}
+
+	for _, host := range c.hostList {
+		if c.host == host {
+			continue
+		}
+
+		if host.lastFailure.After(time.Now().Add(-backoffDuration)) {
+			continue
+		}
+
+		var err error
+		c.host = host
+		c.conn, err = net.DialTimeout("tcp", host.address, connectTimeout)
+		if err != nil {
+			c.markFailure(err)
+			continue
+		}
+
+		err = c.writeNamenodeHandshake()
+		if err != nil {
+			c.markFailure(err)
+			continue
+		}
+
+		break
+	}
+
+	if c.conn == nil {
+		return fmt.Errorf("No available namenodes")
```

Comment:
`No` shouldn't be capitalized. This can also just be `errors.New`. Better, though, would be to include the last error, so that this isn't completely opaque in degenerate cases: something like `fmt.Errorf("no available namenodes: %s", lastErr)`